src/python/pants/backend/experimental/python/lockfile.py

```diff
@@ -104,8 +111,12 @@ async def generate_lockfile(
         return LockGoal(exit_code=0)
 
     input_requirements_get = Get(
-        Digest, CreateDigest([FileContent("requirements.in", "\n".join(reqs).encode())])
+        Digest, CreateDigest([FileContent("requirements.in", "\n".join(reqs.req_strings).encode())])
     )
+    # TODO: Figure out which interpreter constraints to use...Likely get it from the
+    #  transitive closure. When we're doing a single global lockfile, it's fine to do that,
+    #  but we need to figure out how this will work with multiple resolves.
```

Comment:
The same set of `@rules` that requests a resolve will eventually also be triggering the sideeffect of creating the lockfile. It has all of the context needed: exactly which constraints/platforms to use, etc.

But that actually suggests an interesting middle ground. Perhaps what we could do (until that codepath can trigger its own sideeffect via #12014), would be to have the warning/error that we render when a lockfile is stale actually render the _exact_ command to run to regenerate the lockfile. Something like:
```
[ERROR]: The resolve for ${named_resolve} (at $lockfile) was stale. To lock the requirements, please run:
    ./pants lock --interpreter-constraints="${constraints_requested_in_relative_codepath}" $lockfile
```

...and then when we're able to do it automatically via #12014, we can change the default from warn/error to "regenerate".

Reply:
...or it could even specify the entire set of input requirements as an argument...

Reply:
It's not blocking feedback, but this would pretty fundamentally affect the design, and is the reason I didn't immediately shipit: with this design, you don't need a separate `tool-lock` command: instead, when you run pants without a lockfile specified, it will warn/error with the precise command to run the command to generate the lockfile that will make the warning go away.

Reply:
Ah to be clear, I want the `tool-lock` goal to go away and be unified with the `lock` goal. `tool-lock` is only to make forward progress.